### ISSUE TITLE
Logic for new taxdata/puf_finalprep/puf_factors_processing.py

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/taxdata#60]
@andersonfrailey,  Below is my version of a script --- `puf_factors_processing.py` --- that converts blowup factors from the way they are expressed inside the taxdata repo to the way they are expressed in the Tax-Calculator repo.  I think the only thing you will have to change is the `input_filename` because the input `StageIFactors.csv` file will not be in the taxdata/puf_finalprep directory but the taxdata/puf_stage1 directory.  Look at some of the taxcalc/tests code to see how the full path is constructed.  You should be able to produce in taxdata the same taxcalc/puf_factors.csv file as in [Tax-Calculator pull request 1178](https://github.com/open-source-economics/Tax-Calculator/pull/1178).
```
"""
puf_factors_processing.py does final preparation of Stage1 blowup factors by
transforming Stage1Factors.csv into puf_factors.csv
"""
# CODING-STYLE CHECKS:
# pep8 --ignore=E402 puf_factors_processing.py
# pylint --disable=locally-disabled puf_factors_processing.py
# (when importing numpy, add "--extension-pkg-whitelist=numpy" pylint option)

import pandas as pd

# pylint: disable=invalid-name
first_data_year = 2009
input_filename = 'StageIFactors.csv'
output_filename = 'puf_factors.csv'

# read in blowup factors used internally in taxdata repository
data = pd.read_csv(input_filename, index_col='YEAR')

# convert some aggregate factors into per-capita factors
elderly_pop = data['APOPSNR']
data['ASOCSEC'] = data['ASOCSEC'] / elderly_pop
pop = data['APOPN']
data['AWAGE'] = data['AWAGE'] / pop
data['ATXPY'] = data['ATXPY'] / pop
data['ASCHCI'] = data['ASCHCI'] / pop
data['ASCHCL'] = data['ASCHCL'] / pop
data['ASCHF'] = data['ASCHF'] / pop
data['AINTS'] = data['AINTS'] / pop
data['ADIVS'] = data['ADIVS'] / pop
data['ASCHEI'] = data['ASCHEI'] / pop
data['ASCHEL'] = data['ASCHEL'] / pop
data['ACGNS'] = data['ACGNS'] / pop
data['ABOOK'] = data['ABOOK'] / pop
data['AGDPN'] = data['AGDPN'] / pop  # TODO: remove line after Growth redesign

# convert factors into "one plus annual proportion change" format
data = 1.0 + data.pct_change()

# specify first row values because pct_change() leaves first year undefined
# (these values have been transferred from Tax-Calculator records.py)
for var in list(data):
    data[var][first_data_year] = 1.0
data['ACGNS'][first_data_year] = 1.1781
data['ADIVS'][first_data_year] = 1.0606
data['AINTS'][first_data_year] = 1.0357
data['ASCHCI'][first_data_year] = 1.0041
data['ASCHCL'][first_data_year] = 1.1629
data['ASCHEI'][first_data_year] = 1.1089
data['ASCHEL'][first_data_year] = 1.2953
data['AUCOMP'][first_data_year] = 1.0034
data['AWAGE'][first_data_year] = 1.0053

# round converted factors to six decimal digits of accuracy
data = data.round(6)

# delete from data variables not used by Tax-Calculator (TC)
TC_USED_VARS = set(['ABOOK',
                    'ACGNS',
                    'ACPIM',
                    'ACPIU',
                    'ADIVS',
                    'AGDPN',  # TODO: remove this line after Growth redesign
                    'AINTS',
                    'AIPD',
                    'APOPN',  # TODO: remove this line after Growth redesign
                    'ASCHCI',
                    'ASCHCL',
                    'ASCHEI',
                    'ASCHEL',
                    'ASCHF',
                    'ASOCSEC',
                    'ATXPY',
                    'AUCOMP',
                    'AWAGE'])
ALL_VARS = set(list(data))
TC_UNUSED_VARS = ALL_VARS - TC_USED_VARS
data = data.drop(TC_UNUSED_VARS, axis=1)

# write out blowup factors used in Tax-Calculator repository
data.to_csv(output_filename, index='YEAR')
```

**Comment by andersonfrailey** [comment]
@martinholmer, running your code I was able to successfully reproduce `put_factors.csv` in [TaxCalc PR #1178](https://github.com/open-source-economics/Tax-Calculator/pull/1178). I will open up a PR in TaxData to add this script unless you want to wait until the the restructuring of the repo is complete.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> running your code I was able to successfully reproduce puf_factors.csv in TaxCalc PR #1178. I will open up a PR in TaxData to add this script unless you want to wait until the the restructuring of the repo is complete.

Excellent!  Wait on the PR for our discussion of the repo restructuring this afternoon.

**Comment by martinholmer** [comment]
Issue #60 has been resolved by pull request #63.
